[Dx-962] Add FAQ For How To Distribute Analytics Keys Across Redis Clusters (#4018)
TykTechnologies/tyk-docs: tyk-docs/content/developer-support/frequently-asked-questions/how_to_distribute_analytics_to_multiple_redis_shards.md

---

> title: "How To Configure Tyk To Distribute Analytics Keys To Multiple Redis Shards"

Reviewer: As with the other FAQ - the title is the solution, not the problem. As a user, you wouldn't know to search for it or even know it's relevant to you. Please rewrite it to reflect the problem that "enable_multiple_analytics_keys" solves

Author: updated

date: 2024-01-22
tags: ["Analytics", "Distributed Analytics", "Redis", "analytics_config.enable_multiple_analytics_keys" ]

> description: "Explains how to configure the Gateway and MDCB for distributing analytics to multiple Redis shards."

Reviewer: same as the comment above. Include the problem to the sentence.

Author: updated

---

> Tyk distributes analytics to Redis storage. A high volume of analytics traffic can decrease performance, since all analytics keys are stored within one Redis server.

Reviewer: We need to start with the problem. "...Do you have this pattren of CPU for your environment and wonder what could it be?..." then explain one possible reason and then explain the solution

Reviewer: also explain the graph in the image. What do we see in it.

Author: updated

{{< img src="/img/faq/enable-multiple-analytics-keys/redis_single.png" >}}

In Redis, *key sharding* is a term used to describe the practice of distributing data across multiple Redis instances or *shards* based on the keys. This feature is provided by [Redis Cluster](https://redis.io/docs/management/scaling/) and provides horizontal scalability and improved performance. 

Tyk supports configuring this behaviour so that analytics keys are distributed across multiple servers within a Redis cluster.

> {{< img src="/img/faq/enable-multiple-analytics-keys/redis_distributed.png" >}}

Reviewer: sorry if I misunderstood- what do we want to show in the 2nd image

Author: updated

This FAQ explains how to configure Tyk to distribute analytic keys across multiple Redis instances.

> ## How To Distribute Analytics To Multiple Redis Shards

Reviewer: 1. This is exactly like the title...after you change the title should be fine
2. As I was reading, I thought that maybe this section should be in key concepts or maybe something else. worth checking with Rich

To distribute analytics traffic across multiple Redis shards effectively you need to configure the Tyk components to leverage the Redis cluster's sharding capabilities.

> 1. **Enable Redis Cluster Support**: Confirm that the *enable_cluster* configuration option is set to true in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageenable_cluster" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#enable_cluster" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configenable_cluster" >}}) configuration files. This setting activates the Redis cluster support within Tyk components.

Reviewer: ```suggestion
1. **Enable Redis Cluster Support**: Confirm that the `enable_cluster` configuration option is set to true in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageenable_cluster" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#enable_cluster" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configenable_cluster" >}}) configuration files. This setting activates the Redis cluster support within Tyk components.
```
```suggestion
1. **Enable Redis Cluster Support**: Confirm that the *enable_cluster* configuration option is set to true in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageenable_cluster" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#enable_cluster" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configenable_cluster" >}}) configuration files. This setting activates the Redis cluster support within Tyk components.
```

> 1. **Enable Redis Cluster Support**: Confirm that the *enable_cluster* configuration option is set to true in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageenable_cluster" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#enable_cluster" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configenable_cluster" >}}) configuration files. This setting activates the Redis cluster support within Tyk components.

Reviewer: not sure "active" is the right term. redis is a cluster. that's a fact. the setting makes sure tyk knows redis is a cluster and treats it as such when doing CRUD for keys in redis

Author: updated

> 2. **Configure Redis Addresses**: Populate the *addrs* array in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageaddrs" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#redis_addrs" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configaddrs" >}}) configuration files (tyk.conf, tyk_analytics.conf and pump.conf) with the addresses of all Redis cluster nodes. This ensures that the Tyk components can interact with the entire Redis cluster. Please refer to the [configure Redis cluster]({{< ref "tyk-stack/tyk-gateway/configuration/redis-cluster#redis-cluster--tyk-gateway" >}}) guide for further details.

Reviewer: ```suggestion
2. **Configure Redis Addresses**: Populate the `addrs` array in the [Tyk Gateway]({{< ref "tyk-oss-gateway/configuration#storageaddrs" >}}), [Tyk Dashboard]({{< ref "tyk-dashboard/configuration#redis_addrs" >}}) and [Tyk Pump]({{< ref "tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables#analytics_storage_configaddrs" >}}) configuration files (tyk.conf, tyk_analytics.conf and pump.conf) with the addresses of all Redis cluster nodes. This ensures that the Tyk components can interact with the entire Redis cluster. Please refer to the [configure Redis cluster]({{< ref "tyk-stack/tyk-gateway/configuration/redis-cluster#redis-cluster--tyk-gateway" >}}) guide for further details.
```

Reviewer: "Redis Cluster" with capital C as it's the products name

> 3. **Adjust Analytics Configuration**: In the Tyk Gateway configuration (tyk.conf), set [analytics_config.enable_multiple_analytics_keys]({{< ref "tyk-oss-gateway/configuration#analytics_configenable_multiple_analytics_keys" >}}) to true. This option allows Tyk to distribute analytics data across multiple keys, which can be sharded by the Redis cluster. There's a corresponding option for MDCB, named [enable_multiple_analytics_keys]({{< ref "tyk-multi-data-centre/mdcb-configuration-options#enable_multiple_analytics_keys" >}}) if using a data plane and sending analytics to MDCB.

Reviewer: ```suggestion
3. **Adjust Analytics Configuration**: In the Tyk Gateway configuration (tyk.conf), set [analytics_config.enable_multiple_analytics_keys]({{< ref "tyk-oss-gateway/configuration#analytics_configenable_multiple_analytics_keys" >}}) to true. This option allows Tyk to distribute analytics data across Redis nodes, using multiple keys for the analytics. There's a corresponding option for Self Managed MDCB, also named [enable_multiple_analytics_keys]({{< ref "tyk-multi-data-centre/mdcb-configuration-options#enable_multiple_analytics_keys" >}}). Useful only if the gateways in the data plane are configured to send analytics to MDCB.
```

> 3. **Adjust Analytics Configuration**: In the Tyk Gateway configuration (tyk.conf), set [analytics_config.enable_multiple_analytics_keys]({{< ref "tyk-oss-gateway/configuration#analytics_configenable_multiple_analytics_keys" >}}) to true. This option allows Tyk to distribute analytics data across multiple keys, which can be sharded by the Redis cluster. There's a corresponding option for MDCB, named [enable_multiple_analytics_keys]({{< ref "tyk-multi-data-centre/mdcb-configuration-options#enable_multiple_analytics_keys" >}}) if using a data plane and sending analytics to MDCB.

Reviewer: not sure I understand the title "adjust...", it's a configuration of a boolean not fine tunning

Author: updated

4. **Optimise Connection Pool Settings**: Adjust the [optimisation_max_idle]({{< ref "tyk-oss-gateway/configuration#storageoptimisation_max_idle" >}}) and [optimisation_max_active]({{< ref "tyk-oss-gateway/configuration#storageoptimisation_max_active" >}}) settings in the configuration files to ensure that the connection pool can handle the analytics workload without overloading any Redis shard.

> 5. **Monitor and Scale**: Continuously monitor the load on each Redis shard. If load imbalances are detected, you may need to reshard your Redis cluster or scale it to better distribute the keys and handle the traffic.

Author: updated

> 6. **Use a Separate Analytics Store**: For high analytics traffic, you can opt to use a dedicated Redis cluster for analytics by setting [enable_separate_analytics_store]({{< ref "tyk-oss-gateway/configuration#enable_separate_analytics_store" >}}) to true in the Tyk Gateway configuration file (tyk.conf) and specifying the separate Redis cluster configuration in the *analytics_storage* section. Please consult the [separated analytics storage]({{< ref "tyk-stack/tyk-pump/separated-analytics-storage" >}}) guide for an example with Tyk Pump that can equally be applied to Tyk Gateway.

Reviewer: ```suggestion
6. **Use a Separate Analytics Store**: For high analytics traffic, you can opt to use a dedicated *Redis Cluster* for analytics by setting [enable_separate_analytics_store]({{< ref "tyk-oss-gateway/configuration#enable_separate_analytics_store" >}}) to true in the Tyk Gateway configuration file (*tyk.conf*) and specifying the separate Redis cluster configuration in the `analytics_storage` section. Please consult the [separated analytics storage]({{< ref "tyk-stack/tyk-pump/separated-analytics-storage" >}}) guide for an example with *Tyk Pump* that can equally be applied to *Tyk Gateway*.
```

Reviewer: field name - back ticks
name - italic
